File: FindAllAnagramsinaString.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if(s.size()<p.size()){ // prevent overflow
            vector<int>ans;
            return ans;
        }
            
        
        vector<int> sCount(26,0);
        vector<int> pCount(26,0);
        
        for(int i=0;i<p.size();i++)
            pCount[p[i]-'a']++;
        
        for(int i=0;i<p.size()-1;i++)
            sCount[s[i]-'a']++;
    
        vector<int>ans;
        for(int i=p.size()-1; i<s.size(); i++){
            sCount[s[i]-'a']++;
            if(sCount==pCount)
                ans.push_back(i-p.size()+1);
            sCount[s[i-p.size()+1]-'a']--;
        }
    
        return ans;
    }
        
};

File: FindAllAnagramsinaString.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        # from collections import defaultdict
        pCount=defaultdict(int)
        sCount=defaultdict(int)
        for c in p:
            pCount[c]+=1
        for c in s[0:len(p)-1]:
            sCount[c]+=1
            
        ans=[]
        for i in range(len(p)-1,len(s)):
            sCount[s[i]]+=1
            if sCount==pCount:
                ans.append(i-len(p)+1)
            sCount[s[i-len(p)+1]]-=1
            if sCount[s[i-len(p)+1]]==0: # if (abaebab, aba) , remove e's area
                del sCount[s[i-len(p)+1]] 
          
        return ans
